src/object/srv_ec_aggregate.c

```diff
@@ -1419,46 +1419,41 @@ agg_process_holes_ult(void *arg)
 	unsigned int		 k = ec_age2k(entry);
 	unsigned long		 ss = entry->ae_cur_stripe.as_stripenum *
 					k * len;
-	unsigned int		 last_ext_end = 0;
+	unsigned long		 last_ext_end = 0;
 	unsigned int		 ext_cnt = 0;
 	unsigned int		 ext_tot_len = 0;
 	unsigned int		 failed_tgts_cnt = 0;
 	int			 rc = 0;
 
-	D_ALLOC_ARRAY(stripe_ud->asu_recxs,
-		      entry->ae_cur_stripe.as_extent_cnt + 1);
-	if (stripe_ud->asu_recxs == NULL) {
-		rc = -DER_NOMEM;
-		goto out;
-	}
 	/* Process extent list to find what to re-replicate -- build recx array
 	 */
 	d_list_for_each_entry_safe(agg_extent, ext_tmp,
-				   &entry->ae_cur_stripe.as_dextents,
-				   ae_link) {
-		if (agg_extent->ae_hole ||
-		    agg_extent->ae_epoch < entry->ae_par_extent.ape_epoch)
+				   &entry->ae_cur_stripe.as_dextents, ae_link) {
+		if (agg_extent->ae_epoch < entry->ae_par_extent.ape_epoch)
```

Comment:
why you remove original "agg_extent->ae_hole ||" check here? for non-hole ext need to ignore it, no?

Reply:
sorry, seems need "!agg_extent->ae_hole ||"..

Reply:
This code determines what parts of the stripe are not present as replicas or as holes.  It does not consider extents that are older than parity, but it does use the hole records.  It loops over all extents in the stripe (replicas and holes) and finds the ranges to covered.  These ranges are re-replicated.

My original code skipped the holes.  This was a bug.

Reply:
it looks to me it is incorrect for the case that with both punch record, and overwritten partial replica ext.
For example there is a full stripe update [0 - 4M]@epoch10, then user punch [1 - 2M]@epoch15, and overwrite paritial ext [3-4M]@epoch20. for this case should re-replicated [0-1M] + [2 -4M] right?
but above code will generate different result, no?

Reply:
My test case for range punch is currently: write full stripe (4+1), punch first cell at subsequent epoch, overwrite the third cell at a later epoch than the punch.   Test verifies that cells one and four are re-replicated.  I changed the test in my build to match your exact scenario: write full stripe, punch cell 2, overwrite cell 3, and it still passes the test (cells 1 and 4 are re-replicated).

Reply:
Why re-replicate 2-3M?  There is already a valid replica present for this range.

Reply:
ok, I think you are right, need not re-replicate existed replica ext! thanks